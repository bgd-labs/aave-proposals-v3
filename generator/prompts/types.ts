import {PoolIdentifier} from '../types';

export interface GenericPrompt<T extends boolean = boolean> {
  message: string;
  required?: T;
  transform?: (value: string) => string;
  defaultValue?: string;
}

export interface GenericPoolPrompt<T extends boolean = boolean> extends GenericPrompt<T> {
  pool: PoolIdentifier;
}
