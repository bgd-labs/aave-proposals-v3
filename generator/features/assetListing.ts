import {CodeArtifact, FEATURE, FeatureModule, PoolIdentifier} from '../types';
import {fetchBorrowUpdate} from './borrowsUpdates';
import {fetchRateStrategyParamsV3} from './rateUpdates';
import {fetchCollateralUpdate} from './collateralsUpdates';
import {fetchCapsUpdate} from './capsUpdates';
import {Listing, ListingWithCustomImpl, TokenImplementations} from './types';
import {CHAIN_TO_CHAIN_ID, getPoolChain, getExplorerLink} from '../common';
import {getContract, isAddress} from 'viem';
import {confirm} from '@inquirer/prompts';
import {TEST_EXECUTE_PROPOSAL} from '../utils/constants';
import {addressPrompt, translateJsAddressToSol} from '../prompts/addressPrompt';
import {stringPrompt} from '../prompts/stringPrompt';
import {translateJsBoolToSol} from '../prompts/boolPrompt';
import {transformNumberToPercent, translateJsPercentToSol} from '../prompts/percentPrompt';
import {transformNumberToHumanReadable, translateJsNumberToSol} from '../prompts/numberPrompt';
import {CHAIN_ID_CLIENT_MAP} from '@bgd-labs/js-utils';
import {IERC20Detailed_ABI} from '@bgd-labs/aave-address-book/abis';

async function fetchListing(pool: PoolIdentifier): Promise<Listing> {
  const asset = await addressPrompt({
    message: 'Enter the address of the asset you want to list',
    required: true,
  });

  const chain = getPoolChain(pool);
  const erc20 = getContract({
    abi: IERC20Detailed_ABI,
    client: CHAIN_ID_CLIENT_MAP[CHAIN_TO_CHAIN_ID[chain]],
    address: asset,
  });
  let symbol = '';
  try {
    symbol = await erc20.read.symbol();
  } catch (e) {
    console.log('could not fetch the symbol - this is likely an error');
    console.log(e);
  }
  const decimals = await erc20.read.decimals();

  return {
    assetSymbol: await stringPrompt({
      message: 'Enter the asset symbol',
      required: true,
      defaultValue: symbol,
    }),
    decimals,
    priceFeed: await addressPrompt({message: 'PriceFeed address', required: true}),
    ...(await fetchCollateralUpdate(pool, true)),
    ...(await fetchBorrowUpdate(true)),
    ...(await fetchCapsUpdate(true)),
    rateStrategyParams: await fetchRateStrategyParamsV3(true),
    asset,
    admin: await addressPrompt({message: 'Emission admin address (optional)', required: false}),
  };
}

async function fetchCustomImpl(): Promise<TokenImplementations> {
  return {
    aToken: await addressPrompt({message: 'aToken implementation', required: true}),
    vToken: await addressPrompt({message: 'vToken implementation', required: true}),
    sToken: await addressPrompt({message: 'sToken implementation', required: true}),
  };
}

function generateAssetListingSol(cfg: Listing) {
  return `asset: ${cfg.assetSymbol},
  assetSymbol: "${cfg.assetSymbol}",
  priceFeed: ${translateJsAddressToSol(cfg.priceFeed)},
  enabledToBorrow: ${translateJsBoolToSol(cfg.enabledToBorrow)},
  borrowableInIsolation: ${translateJsBoolToSol(cfg.borrowableInIsolation)},
  withSiloedBorrowing: ${translateJsBoolToSol(cfg.withSiloedBorrowing)},
  flashloanable: ${translateJsBoolToSol(cfg.flashloanable)},
  ltv: ${translateJsPercentToSol(cfg.ltv)},
  liqThreshold: ${translateJsPercentToSol(cfg.liqThreshold)},
  liqBonus: ${translateJsPercentToSol(cfg.liqBonus)},
  reserveFactor: ${translateJsPercentToSol(cfg.reserveFactor)},
  supplyCap: ${translateJsNumberToSol(cfg.supplyCap)},
  borrowCap: ${translateJsNumberToSol(cfg.borrowCap)},
  debtCeiling: ${translateJsNumberToSol(cfg.debtCeiling)},
  liqProtocolFee: ${translateJsPercentToSol(cfg.liqProtocolFee)},
  rateStrategyParams: IAaveV3ConfigEngine.InterestRateInputData({
     optimalUsageRatio: ${translateJsPercentToSol(cfg.rateStrategyParams.optimalUtilizationRate)},
     baseVariableBorrowRate: ${translateJsPercentToSol(
       cfg.rateStrategyParams.baseVariableBorrowRate,
     )},
     variableRateSlope1: ${translateJsPercentToSol(cfg.rateStrategyParams.variableRateSlope1)},
     variableRateSlope2: ${translateJsPercentToSol(cfg.rateStrategyParams.variableRateSlope2)}
  })`;
}

export const assetListing: FeatureModule<Listing[]> = {
  value: FEATURE.ASSET_LISTING,
  description: 'newListings (listing a new asset)',
  async cli({pool}) {
    const response: Listing[] = [];
    console.log(`Fetching information for Assets assets on ${pool}`);
    let more: boolean = true;
    while (more) {
      response.push(await fetchListing(pool));
      more = await confirm({message: 'Do you want to list another asset?', default: false});
    }
    return response;
  },
  build({pool, cfg}) {
    const response: CodeArtifact = {
      code: {
        constants: cfg.map((cfg) => {
          let listingConstant = `address public constant ${cfg.assetSymbol} = ${translateJsAddressToSol(cfg.asset)};\n`;
          listingConstant += `uint256 public constant ${cfg.assetSymbol}_SEED_AMOUNT = 1e${cfg.decimals};\n`;
          if (isAddress(cfg.admin)) {
            listingConstant += `address public constant ${cfg.assetSymbol}_LM_ADMIN = ${translateJsAddressToSol(cfg.admin)};\n`;
          }
          return listingConstant;
        }),
        execute: cfg.map((cfg) => {
          let listingExe = `IERC20(${cfg.assetSymbol}).forceApprove(address(${pool}.POOL), ${cfg.assetSymbol}_SEED_AMOUNT);\n`;
          listingExe += `${pool}.POOL.supply(${cfg.assetSymbol}, ${cfg.assetSymbol}_SEED_AMOUNT, address(${pool}.COLLECTOR), 0);\n`;
          if (isAddress(cfg.admin)) {
            listingExe += `\n(address a${cfg.assetSymbol}, , ) = ${pool}.AAVE_PROTOCOL_DATA_PROVIDER.getReserveTokensAddresses(${cfg.assetSymbol});\n`;
            listingExe += `IEmissionManager(${pool}.EMISSION_MANAGER).setEmissionAdmin(${cfg.assetSymbol}, ${cfg.assetSymbol}_ADMIN);\n`;
            listingExe += `IEmissionManager(${pool}.EMISSION_MANAGER).setEmissionAdmin(a${cfg.assetSymbol}, ${cfg.assetSymbol}_ADMIN);\n`;
          }
          return listingExe;
        }),
        fn: [
          `function newListings() public pure override returns (IAaveV3ConfigEngine.Listing[] memory) {
          IAaveV3ConfigEngine.Listing[] memory listings = new IAaveV3ConfigEngine.Listing[](${
            cfg.length
          });

          ${cfg
            .map(
              (cfg, ix) => `listings[${ix}] = IAaveV3ConfigEngine.Listing({
               ${generateAssetListingSol(cfg)}
             });`,
            )
            .join('\n')}

          return listings;
        }`,
        ],
      },
      test: {
        fn: cfg.map((cfg) => {
          let listingTest = `function test_collectorHas${cfg.assetSymbol}Funds() public {
            ${TEST_EXECUTE_PROPOSAL}
            (address aTokenAddress, , ) = ${pool}.AAVE_PROTOCOL_DATA_PROVIDER.getReserveTokensAddresses(proposal.${cfg.assetSymbol}());
            assertGe(IERC20(aTokenAddress).balanceOf(address(${pool}.COLLECTOR)), 10 ** ${cfg.decimals});
          }\n`;
          if (isAddress(cfg.admin)) {
            listingTest += `\nfunction test_${cfg.assetSymbol}Admin() public {
	      ${TEST_EXECUTE_PROPOSAL}
              (address a${cfg.assetSymbol}, , ) = ${pool}.AAVE_PROTOCOL_DATA_PROVIDER.getReserveTokensAddresses(proposal.${cfg.assetSymbol}());
	      assertEq(IEmissionManager(${pool}.EMISSION_MANAGER).getEmissionAdmin(proposal.${cfg.assetSymbol}()), proposal.${cfg.assetSymbol}_ADMIN());
	      assertEq(IEmissionManager(${pool}.EMISSION_MANAGER).getEmissionAdmin(a${cfg.assetSymbol}), proposal.${cfg.assetSymbol}_ADMIN());
	    }\n`;
          }
          return listingTest;
        }),
      },
      aip: {
        specification: cfg.map((cfg) => {
          let listingTemplate = `The table below illustrates the configured risk parameters for **${cfg.assetSymbol}**\n\n`;
          listingTemplate += `| Parameter | Value |\n`;
          listingTemplate += `| --- | --: |\n`;
          listingTemplate += `| Isolation Mode | ${cfg.debtCeiling !== '0'} |\n`;
          listingTemplate += `| Borrowable | ${cfg.enabledToBorrow} |\n`;
          listingTemplate += `| Collateral Enabled | ${!!cfg.liqThreshold} |\n`;
          listingTemplate += `| Supply Cap (${cfg.assetSymbol}) | ${transformNumberToHumanReadable(
            cfg.supplyCap,
          )} |\n`;
          listingTemplate += `| Borrow Cap (${cfg.assetSymbol}) | ${transformNumberToHumanReadable(
            cfg.borrowCap,
          )} |\n`;
          listingTemplate += `| Debt Ceiling | USD ${transformNumberToHumanReadable(
            cfg.debtCeiling,
          )} |\n`;
          listingTemplate += `| LTV | ${transformNumberToPercent(cfg.ltv)} |\n`;
          listingTemplate += `| LT | ${transformNumberToPercent(cfg.liqThreshold)} |\n`;
          listingTemplate += `| Liquidation Bonus	| ${transformNumberToPercent(cfg.liqBonus)} |\n`;
          listingTemplate += `| Liquidation Protocol Fee | ${transformNumberToPercent(
            cfg.liqProtocolFee,
          )} |\n`;
          listingTemplate += `| Reserve Factor | ${transformNumberToPercent(
            cfg.reserveFactor,
          )} |\n`;
          listingTemplate += `| Base Variable Borrow Rate	| ${transformNumberToPercent(
            cfg.rateStrategyParams.baseVariableBorrowRate,
          )} |\n`;
          listingTemplate += `| Variable Slope 1 | ${transformNumberToPercent(
            cfg.rateStrategyParams.variableRateSlope1,
          )} |\n`;
          listingTemplate += `| Variable Slope 2 | ${transformNumberToPercent(
            cfg.rateStrategyParams.variableRateSlope2,
          )} |\n`;
          listingTemplate += `| Uoptimal | ${transformNumberToPercent(
            cfg.rateStrategyParams.optimalUtilizationRate,
          )} |\n`;
          listingTemplate += `| Flashloanable	| ${cfg.flashloanable} |\n`;
          listingTemplate += `| Siloed Borrowing	| ${cfg.withSiloedBorrowing} |\n`;
          listingTemplate += `| Borrowable in Isolation | ${cfg.borrowableInIsolation} |\n`;
          listingTemplate += `| Oracle | ${cfg.priceFeed} |\n`;
          if (isAddress(cfg.admin)) {
            listingTemplate += `\nAdditionaly [${cfg.admin}](${getExplorerLink(CHAIN_TO_CHAIN_ID[getPoolChain(pool)], cfg.admin)}) has been set as the emission admin for ${cfg.assetSymbol} and the corresponding aToken.\n`;
          }
          return listingTemplate;
        }),
      },
    };
    return response;
  },
};

export const assetListingCustom: FeatureModule<ListingWithCustomImpl[]> = {
  value: FEATURE.ASSET_LISTING_CUSTOM,
  description: 'newListingsCustom (listing a new asset, with custom implementations)',
  async cli({pool}) {
    const response: ListingWithCustomImpl[] = [];
    let more: boolean = true;
    while (more) {
      response.push({base: await fetchListing(pool), implementations: await fetchCustomImpl()});
      more = await confirm({message: 'Do you want to list another asset?', default: false});
    }
    return response;
  },
  build({pool, cfg}) {
    const response: CodeArtifact = {
      code: {
        constants: cfg.map(
          (cfg) =>
            `address public constant ${cfg.base.assetSymbol} = ${translateJsAddressToSol(
              cfg.base.asset,
            )};`,
        ),
        execute: cfg.map(
          (cfg) =>
            `IERC20(${cfg.base.assetSymbol}).forceApprove(address(${pool}.POOL), 10 ** ${cfg.base.decimals});
            ${pool}.POOL.supply(${cfg.base.assetSymbol}, 10 ** ${cfg.base.decimals}, ${pool}.COLLECTOR, 0);`,
        ),
        fn: [
          `function newListingsCustom() public pure override returns (IAaveV3ConfigEngine.ListingWithCustomImpl[] memory) {
          IAaveV3ConfigEngine.ListingWithCustomImpl[] memory listings = new IAaveV3ConfigEngine.ListingWithCustomImpl[](${
            cfg.length
          });

          ${cfg
            .map(
              (cfg, ix) => `listings[${ix}] = IAaveV3ConfigEngine.ListingWithCustomImpl(
                IAaveV3ConfigEngine.Listing({
                  ${generateAssetListingSol(cfg.base)},
                  IAaveV3ConfigEngine.TokenImplementations({
                    aToken: ${translateJsAddressToSol(cfg.implementations.aToken)},
                    vToken: ${translateJsAddressToSol(cfg.implementations.vToken)},
                    sToken: ${translateJsAddressToSol(cfg.implementations.sToken)}
                  })
                })
             );`,
            )
            .join('\n')}

          return listings;
        }`,
        ],
      },
      test: {
        fn: cfg.map(
          (cfg) => `function test_collectorHas${cfg.base.assetSymbol}Funds() public {
            ${TEST_EXECUTE_PROPOSAL}
            (address aTokenAddress, , ) = ${pool}.AAVE_PROTOCOL_DATA_PROVIDER.getReserveTokensAddresses(proposal.${cfg.base.assetSymbol}());
            assertGte(IERC20(aTokenAddress).balanceOf(${pool}.COLLECTOR), 10 ** ${cfg.base.decimals});
          }`,
        ),
      },
    };
    return response;
  },
};
