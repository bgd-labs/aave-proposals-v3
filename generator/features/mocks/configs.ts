import {Options} from '../../types';
import {EModeCategoryUpdate, Listing, PriceFeedUpdate} from '../types';

export const MOCK_OPTIONS: Options = {
  pools: ['AaveV3Ethereum'],
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  author: 'test',
  discussion: 'test',
  snapshot: 'test',
};

export const assetListingConfig: Listing[] = [
  {
    assetSymbol: 'PSP',
    decimals: 18,
    priceFeed: '0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8',
    ltv: '40_00',
    liqThreshold: '50_00',
    liqBonus: '5_00',
    debtCeiling: '100_000',
    liqProtocolFee: '20_00',
    enabledToBorrow: 'EngineFlags.ENABLED',
    flashloanable: 'EngineFlags.ENABLED',
    stableRateModeEnabled: 'EngineFlags.DISABLED',
    borrowableInIsolation: 'EngineFlags.DISABLED',
    withSiloedBorrowing: 'EngineFlags.DISABLED',
    reserveFactor: '20_00',
    supplyCap: '10_000',
    borrowCap: '5_000',
    rateStrategyParams: {
      optimalUtilizationRate: '_bpsToRay(80_00)',
      baseVariableBorrowRate: '_bpsToRay(0)',
      variableRateSlope1: '_bpsToRay(10_00)',
      variableRateSlope2: '_bpsToRay(100_00)',
      stableRateSlope1: '_bpsToRay(10_00)',
      stableRateSlope2: '_bpsToRay(100_00)',
      baseStableRateOffset: '_bpsToRay(1_00)',
      stableRateExcessOffset: '_bpsToRay(0)',
      optimalStableToTotalDebtRatio: '_bpsToRay(10_00)',
    },
    eModeCategory: 'AaveV3EthereumEModes.NONE',
    asset: '0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5',
  },
];

export const priceFeedsUpdateConfig: PriceFeedUpdate[] = [
  {
    asset: 'DAI',
    priceFeed: '0xae7ab96520de3a18e5e111b5eaab095312d7fe84',
  },
];

export const emodeUpdates: EModeCategoryUpdate[] = [
  {
    eModeCategory: 2,
    ltv: '20_00',
    liqThreshold: '30_00',
    liqBonus: '5_00',
    priceSource: '0x0000000000000000000000000000000000000000',
    label: 'label',
  },
  {
    eModeCategory: 'AaveV3EthereumEModes.ETH_CORRELATED',
    ltv: 'EngineFlags.KEEP_CURRENT',
    liqThreshold: '50_00',
    liqBonus: 'EngineFlags.KEEP_CURRENT',
    priceSource: '',
    label: '',
  },
];
