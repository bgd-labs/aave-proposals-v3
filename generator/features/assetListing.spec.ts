// sum.test.js
import {expect, describe, it} from 'vitest';
import {assetListing} from './assetListing';
import {MOCK_OPTIONS, assetListingConfig} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';

describe('feature: assetListing', () => {
  it('should return reasonable code', () => {
    const output = assetListing.build({
      options: MOCK_OPTIONS,
      pool: 'AaveV3Ethereum',
      cfg: assetListingConfig,
      cache: {blockNumber: 42},
    });
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      [MOCK_OPTIONS.pools[0]]: {
        pool: MOCK_OPTIONS.pools[0],
        artifacts: [
          assetListing.build({
            options: MOCK_OPTIONS,
            pool: 'AaveV3Ethereum',
            cfg: assetListingConfig,
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.ASSET_LISTING]: assetListingConfig},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles(MOCK_OPTIONS, poolConfigs);
    expect(files).toMatchSnapshot();
  });

  it('regression: isolation mode should be flase when ceiling is zero', async () => {
    const poolConfigs: PoolConfigs = {
      [MOCK_OPTIONS.pools[0]]: {
        pool: MOCK_OPTIONS.pools[0],
        artifacts: [
          assetListing.build({
            options: MOCK_OPTIONS,
            pool: 'AaveV3Ethereum',
            cfg: [{...assetListingConfig[0], debtCeiling: '0'}],
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.ASSET_LISTING]: [{...assetListingConfig[0], debtCeiling: 0}]},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles(MOCK_OPTIONS, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
