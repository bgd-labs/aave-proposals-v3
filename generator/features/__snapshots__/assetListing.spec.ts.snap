// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: assetListing > should properly generate files 1`] = `
{
  "aip": "---
title: \\"test\\"
author: \\"test\\"
discussions: \\"test\\"
---

## Simple Summary

## Motivation

## Specification

## References

- Implementation: [AaveV3Ethereum](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV3Ethereum_Test/AaveV3Ethereum_Test_20231023.sol)
- Tests: [AaveV3Ethereum](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV3Ethereum_Test/AaveV3Ethereum_Test_20231023.t.sol)
- [Snapshot](test)
- [Discussion](test)

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
",
  "jsonConfig": "{
  \\"rootOptions\\": {
    \\"pools\\": [
      \\"AaveV3Ethereum\\"
    ],
    \\"title\\": \\"test\\",
    \\"shortName\\": \\"Test\\",
    \\"date\\": \\"20231023\\",
    \\"author\\": \\"test\\",
    \\"discussion\\": \\"test\\",
    \\"snapshot\\": \\"test\\"
  },
  \\"poolOptions\\": {
    \\"AaveV3Ethereum\\": {
      \\"configs\\": {
        \\"ASSET_LISTING\\": [
          {
            \\"assetSymbol\\": \\"PSP\\",
            \\"decimals\\": 18,
            \\"priceFeed\\": \\"0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8\\",
            \\"ltv\\": \\"40_00\\",
            \\"liqThreshold\\": \\"50_00\\",
            \\"liqBonus\\": \\"5_00\\",
            \\"debtCeiling\\": \\"100_000\\",
            \\"liqProtocolFee\\": \\"20_00\\",
            \\"enabledToBorrow\\": \\"EngineFlags.ENABLED\\",
            \\"flashloanable\\": \\"EngineFlags.ENABLED\\",
            \\"stableRateModeEnabled\\": \\"EngineFlags.DISABLED\\",
            \\"borrowableInIsolation\\": \\"EngineFlags.DISABLED\\",
            \\"withSiloedBorrowing\\": \\"EngineFlags.DISABLED\\",
            \\"reserveFactor\\": \\"20_00\\",
            \\"supplyCap\\": \\"10_000\\",
            \\"borrowCap\\": \\"5_000\\",
            \\"rateStrategyParams\\": {
              \\"optimalUtilizationRate\\": \\"_bpsToRay(80_00)\\",
              \\"baseVariableBorrowRate\\": \\"_bpsToRay(0)\\",
              \\"variableRateSlope1\\": \\"_bpsToRay(10_00)\\",
              \\"variableRateSlope2\\": \\"_bpsToRay(100_00)\\",
              \\"stableRateSlope1\\": \\"_bpsToRay(10_00)\\",
              \\"stableRateSlope2\\": \\"_bpsToRay(100_00)\\",
              \\"baseStableRateOffset\\": \\"_bpsToRay(1_00)\\",
              \\"stableRateExcessOffset\\": \\"_bpsToRay(0)\\",
              \\"optimalStableToTotalDebtRatio\\": \\"_bpsToRay(10_00)\\"
            },
            \\"eModeCategory\\": \\"AaveV3EthereumEModes.NONE\\",
            \\"asset\\": \\"0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5\\"
          }
        ]
      },
      \\"features\\": [
        \\"ASSET_LISTING\\"
      ]
    }
  }
}",
  "payloads": [
    {
      "contractName": "AaveV3Ethereum_Test_20231023",
      "payload": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV3Ethereum, AaveV3EthereumEModes} from 'aave-address-book/AaveV3Ethereum.sol';
import {AaveV3PayloadEthereum} from 'aave-helpers/v3-config-engine/AaveV3PayloadEthereum.sol';
import {EngineFlags} from 'aave-helpers/v3-config-engine/EngineFlags.sol';
import {IAaveV3ConfigEngine} from 'aave-helpers/v3-config-engine/IAaveV3ConfigEngine.sol';
import {IV3RateStrategyFactory} from 'aave-helpers/v3-config-engine/IV3RateStrategyFactory.sol';
import {IERC20} from 'solidity-utils/contracts/oz-common/interfaces/IERC20.sol';
import {SafeERC20} from 'solidity-utils/contracts/oz-common/SafeERC20.sol';

/**
 * @title test
 * @author test
 * - Snapshot: test
 * - Discussion: test
 */
contract AaveV3Ethereum_Test_20231023 is AaveV3PayloadEthereum {
  using SafeERC20 for IERC20;

  address public constant PSP = 0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5;
  uint256 internal constant PSP_SEED_AMOUNT = 10 ** 18;

  function _postExecute() internal override {
    IERC20(PSP).forceApprove(address(AaveV3Ethereum.POOL), PSP_SEED_AMOUNT);
    AaveV3Ethereum.POOL.supply(PSP, PSP_SEED_AMOUNT, address(AaveV3Ethereum.COLLECTOR), 0);
  }

  function newListings() public pure override returns (IAaveV3ConfigEngine.Listing[] memory) {
    IAaveV3ConfigEngine.Listing[] memory listings = new IAaveV3ConfigEngine.Listing[](1);

    listings[0] = IAaveV3ConfigEngine.Listing({
      asset: PSP,
      assetSymbol: 'PSP',
      priceFeed: 0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8,
      eModeCategory: AaveV3EthereumEModes.NONE,
      enabledToBorrow: EngineFlags.ENABLED,
      stableRateModeEnabled: EngineFlags.DISABLED,
      borrowableInIsolation: EngineFlags.DISABLED,
      withSiloedBorrowing: EngineFlags.DISABLED,
      flashloanable: EngineFlags.ENABLED,
      ltv: 40_00,
      liqThreshold: 50_00,
      liqBonus: 5_00,
      reserveFactor: 20_00,
      supplyCap: 10_000,
      borrowCap: 5_000,
      debtCeiling: 100_000,
      liqProtocolFee: 20_00,
      rateStrategyParams: IV3RateStrategyFactory.RateStrategyParams({
        optimalUsageRatio: _bpsToRay(80_00),
        baseVariableBorrowRate: _bpsToRay(0),
        variableRateSlope1: _bpsToRay(10_00),
        variableRateSlope2: _bpsToRay(100_00),
        stableRateSlope1: _bpsToRay(10_00),
        stableRateSlope2: _bpsToRay(100_00),
        baseStableRateOffset: _bpsToRay(1_00),
        stableRateExcessOffset: _bpsToRay(0),
        optimalStableToTotalDebtRatio: _bpsToRay(10_00)
      })
    });

    return listings;
  }
}
",
      "test": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {GovV3Helpers} from 'aave-helpers/GovV3Helpers.sol';
import {AaveV3Ethereum} from 'aave-address-book/AaveV3Ethereum.sol';
import {IERC20} from 'solidity-utils/contracts/oz-common/interfaces/IERC20.sol';

import 'forge-std/Test.sol';
import {ProtocolV3TestBase, ReserveConfig} from 'aave-helpers/ProtocolV3TestBase.sol';
import {AaveV3Ethereum_Test_20231023} from './AaveV3Ethereum_Test_20231023.sol';

/**
 * @dev Test for AaveV3Ethereum_Test_20231023
 * command: make test-contract filter=AaveV3Ethereum_Test_20231023
 */
contract AaveV3Ethereum_Test_20231023_Test is ProtocolV3TestBase {
  AaveV3Ethereum_Test_20231023 internal proposal;

  function setUp() public {
    vm.createSelectFork(vm.rpcUrl('mainnet'), 18487480);
    proposal = new AaveV3Ethereum_Test_20231023();
  }

  /**
   * @dev executes the generic test suite including e2e and config snapshots
   */
  function test_defaultProposalExecution() public {
    defaultTest('AaveV3Ethereum_Test_20231023', AaveV3Ethereum.POOL, address(proposal));
  }

  function test_collectorHasPSPFunds() public {
    GovV3Helpers.executePayload(vm, address(proposal));
    (address aTokenAddress, , ) = AaveV3Ethereum
      .AAVE_PROTOCOL_DATA_PROVIDER
      .getReserveTokensAddresses(proposal.PSP());
    assertGe(IERC20(aTokenAddress).balanceOf(address(AaveV3Ethereum.COLLECTOR)), 10 ** 18);
  }
}
",
    },
  ],
  "script": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {GovV3Helpers, IPayloadsControllerCore, PayloadsControllerUtils} from 'aave-helpers/GovV3Helpers.sol';
import {EthereumScript} from 'aave-helpers/ScriptUtils.sol';
import {AaveV3Ethereum_Test_20231023} from './AaveV3Ethereum_Test_20231023.sol';

/**
 * @dev Deploy Ethereum
 * command: make deploy-ledger contract=src/20231023_AaveV3Ethereum_Test/Test_20231023.s.sol:DeployEthereum chain=mainnet
 */
contract DeployEthereum is EthereumScript {
  function run() external broadcast {
    // deploy payloads
    AaveV3Ethereum_Test_20231023 payload0 = new AaveV3Ethereum_Test_20231023();

    // compose action
    IPayloadsControllerCore.ExecutionAction[]
      memory actions = new IPayloadsControllerCore.ExecutionAction[](1);
    actions[0] = GovV3Helpers.buildAction(address(payload0));

    // register action at payloadsController
    GovV3Helpers.createPayload(actions);
  }
}

/**
 * @dev Create Proposal
 * command: make deploy-ledger contract=src/20231023_AaveV3Ethereum_Test/Test_20231023.s.sol:CreateProposal chain=mainnet
 */
contract CreateProposal is EthereumScript {
  function run() external {
    // create payloads
    PayloadsControllerUtils.Payload[] memory payloads = new PayloadsControllerUtils.Payload[](1);

    // compose actions for validation
    IPayloadsControllerCore.ExecutionAction[]
      memory actionsEthereum = new IPayloadsControllerCore.ExecutionAction[](1);
    actionsEthereum[0] = GovV3Helpers.buildAction(address(0));
    payloads[0] = GovV3Helpers.buildMainnetPayload(vm, actionsEthereum);

    // create proposal
    vm.startBroadcast();
    GovV3Helpers.createProposal2_5(
      payloads,
      GovV3Helpers.ipfsHashFile(vm, 'src/20231023_AaveV3Ethereum_Test/Test.md')
    );
  }
}
",
}
`;

exports[`feature: assetListing > should return reasonable code 1`] = `
{
  "code": {
    "constants": [
      "address public constant PSP = 0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5;",
      "uint256 internal constant PSP_SEED_AMOUNT = 10 ** 18;",
    ],
    "execute": [
      "IERC20(PSP).forceApprove(address(AaveV3Ethereum.POOL), PSP_SEED_AMOUNT);
            AaveV3Ethereum.POOL.supply(PSP, PSP_SEED_AMOUNT, address(AaveV3Ethereum.COLLECTOR), 0);",
    ],
    "fn": [
      "function newListings() public pure override returns (IAaveV3ConfigEngine.Listing[] memory) {
          IAaveV3ConfigEngine.Listing[] memory listings = new IAaveV3ConfigEngine.Listing[](1);

          listings[0] = IAaveV3ConfigEngine.Listing({
               asset: PSP,
               assetSymbol: \\"PSP\\",
               priceFeed: 0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8,
               eModeCategory: AaveV3EthereumEModes.NONE,
               enabledToBorrow: EngineFlags.ENABLED,
               stableRateModeEnabled: EngineFlags.DISABLED,
               borrowableInIsolation: EngineFlags.DISABLED,
               withSiloedBorrowing: EngineFlags.DISABLED,
               flashloanable: EngineFlags.ENABLED,
               ltv: 40_00,
               liqThreshold: 50_00,
               liqBonus: 5_00,
               reserveFactor: 20_00,
               supplyCap: 10_000,
               borrowCap: 5_000,
               debtCeiling: 100_000,
               liqProtocolFee: 20_00,
               rateStrategyParams: IV3RateStrategyFactory.RateStrategyParams({
                  optimalUsageRatio: _bpsToRay(80_00),
                  baseVariableBorrowRate: _bpsToRay(0),
                  variableRateSlope1: _bpsToRay(10_00),
                  variableRateSlope2: _bpsToRay(100_00),
                  stableRateSlope1: _bpsToRay(10_00),
                  stableRateSlope2: _bpsToRay(100_00),
                  baseStableRateOffset: _bpsToRay(1_00),
                  stableRateExcessOffset: _bpsToRay(0),
                  optimalStableToTotalDebtRatio: _bpsToRay(10_00)
              })
             });

          return listings;
        }",
    ],
  },
  "test": {
    "fn": [
      "function test_collectorHasPSPFunds() public {
            GovV3Helpers.executePayload(vm,address(proposal));
            (address aTokenAddress, , ) = AaveV3Ethereum.AAVE_PROTOCOL_DATA_PROVIDER.getReserveTokensAddresses(proposal.PSP());
            assertGe(IERC20(aTokenAddress).balanceOf(address(AaveV3Ethereum.COLLECTOR)), 10 ** 18);
          }",
    ],
  },
}
`;
